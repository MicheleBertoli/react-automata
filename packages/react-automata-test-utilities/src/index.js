/* eslint-disable import/prefer-default-export */

export { default as testStateMachine } from './testStateMachine'
