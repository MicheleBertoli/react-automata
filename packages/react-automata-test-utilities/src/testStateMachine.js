import idx from 'idx'
import React from 'react'
import TestRenderer from 'react-test-renderer'
import { getShortestPaths } from 'xstate/lib/graph'
import { stringify } from 'react-automata-utilities'

const testStateMachine = (Component, options = {}) => {
  const paths = getShortestPaths(Component.machine, options.extendedState)

  Object.keys(paths).forEach(key => {
    const initialData = idx(options, _ => _.fixtures.initialData)
    const renderer = TestRenderer.create(
      <Component initialData={initialData} />
    )
    const instance = renderer.getInstance()

    paths[key].forEach(({ event, state }) => {
      const fixtures = idx(options, _ => _.fixtures[state][event])

      instance.handleTransition(event, fixtures)
    })

    const machineState =
      instance.state.machineState.toString() ||
      stringify(instance.state.machineState.value).join(',')

    expect(renderer.toJSON()).toMatchSnapshot(machineState)
  })
}

export default testStateMachine
