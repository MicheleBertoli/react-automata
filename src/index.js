export { default as Action } from './Action'
export { default as State } from './State'
export { default as testStateMachine } from './testStateMachine'
export { default as withStateMachine } from './withStateMachine'
