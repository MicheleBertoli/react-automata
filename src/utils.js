import globToRegExp from 'glob-to-regexp'
import idx from 'idx'

export const DEFAULT_CHANNEL = 'DEFAULT'

export const getComponentName = Component =>
  Component.displayName || Component.name || 'Component'

export const isStateless = Component =>
  !idx(Component, _ => _.prototype.isReactComponent)

export const stringify = (state, path = []) => {
  if (typeof state === 'string') {
    return path.concat(state).join('.')
  }

  return Object.keys(state).reduce(
    (prev, key) => prev.concat(stringify(state[key], path.concat(key))),
    []
  )
}

export const getPatterns = glob =>
  Array.isArray(glob)
    ? glob.map(pattern => globToRegExp(pattern))
    : [globToRegExp(glob)]

export const matches = (patterns, value) => {
  const values = Array.isArray(value) ? value : [value]

  return patterns.some(pattern => values.some(val => pattern.test(val)))
}
