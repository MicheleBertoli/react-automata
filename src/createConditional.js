import idx from 'idx'
import memoize from 'memoize-one'
import PropTypes from 'prop-types'
import React from 'react'
import Context from './context'
import { DEFAULT_CHANNEL, getPatterns, matches } from './utils'

const createConditional = (displayName, contextField) => {
  class Conditional extends React.Component {
    componentDidMount() {
      if (this.isVisible && this.props.onShow) {
        this.props.onShow()
      }
    }

    componentDidUpdate() {
      if (!this.wasVisible && this.isVisible && this.props.onShow) {
        this.props.onShow()
      }

      if (this.wasVisible && !this.isVisible && this.props.onHide) {
        this.props.onHide()
      }
    }

    getPatterns = memoize(getPatterns)

    matches = memoize(matches)

    render() {
      this.wasVisible = this.isVisible
      const patterns = this.getPatterns(this.props.is)
      this.isVisible = this.matches(patterns, this.props.value)

      if (this.props.render) {
        return this.props.render(this.isVisible)
      }

      return this.isVisible ? this.props.children : null
    }
  }

  const Container = props => (
    <Context.Consumer>
      {context => (
        <Conditional
          {...props}
          value={idx(
            context,
            _ => _[props.channel || DEFAULT_CHANNEL][contextField]
          )}
        />
      )}
    </Context.Consumer>
  )

  Container.defaultProps = {
    children: null,
  }

  Container.displayName = displayName

  Container.propTypes = {
    channel: PropTypes.string,
    children: PropTypes.node,
    is: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.string),
      PropTypes.string,
    ]),
    render: PropTypes.func,
    onHide: PropTypes.func,
    onShow: PropTypes.func,
  }

  return Container
}

export default createConditional
