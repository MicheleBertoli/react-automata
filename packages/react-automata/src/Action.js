import createConditional from './createConditional'

export default createConditional('Action', 'actions')
