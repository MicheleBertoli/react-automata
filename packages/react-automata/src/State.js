import createConditional from './createConditional'

export default createConditional('State', 'machineState')
