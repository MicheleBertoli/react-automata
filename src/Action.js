import PropTypes from 'prop-types'
import createConditional from './createConditional'

const displayName = 'Action'

const propTypes = {
  hide: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
  show: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
}

const matches = (value, actions) =>
  Array.isArray(value)
    ? actions.some(action => value.includes(action))
    : actions.includes(value)

export const shouldShow = (props, context) =>
  matches(props.show, context.actions)

export const shouldHide = (props, context) =>
  props.hide
    ? matches(props.hide, context.actions)
    : !matches(props.show, context.actions)

export default createConditional({
  displayName,
  propTypes: process.env.NODE_ENV !== 'production' ? propTypes : null,
  shouldShow,
  shouldHide,
})
