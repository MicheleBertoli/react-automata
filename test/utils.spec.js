import React from 'react'
import { getPatterns, isStateless, matches, stringify } from '../src/utils'

describe('isStateless', () => {
  test('true', () => {
    const Component = () => <div />
    const result = isStateless(Component)

    expect(result).toBe(true)
  })

  test('false', () => {
    class Component extends React.Component {
      render() {
        return <div />
      }
    }
    const result = isStateless(Component)

    expect(result).toBe(false)
  })
})

describe('stringify', () => {
  test('true', () => {
    const state = {
      a: {
        b: {
          c: 'd',
          e: 'f',
        },
        g: 'h',
      },
      b: 'c',
    }
    const result = stringify(state)
    const expected = ['a.b.c.d', 'a.b.e.f', 'a.g.h', 'b.c']

    expect(result).toEqual(expected)
  })
})

test('getPatterns', () => {
  const patterns1 = getPatterns('foo')
  const patterns2 = getPatterns(['foo'])

  expect(patterns1).toEqual(expect.arrayContaining(patterns2))
  expect(patterns1).toMatchSnapshot()
})

test('matches', () => {
  const patterns = [/^foo$/]

  expect(matches(patterns, 'foo')).toBe(true)
  expect(matches(patterns, ['foo'])).toBe(true)
  expect(matches(patterns, 'bar')).toBe(false)
  expect(matches(patterns, ['bar'])).toBe(false)
})
