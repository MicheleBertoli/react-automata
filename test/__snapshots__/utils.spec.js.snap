// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPatterns 1`] = `
Array [
  /\\^foo\\$/,
]
`;
